### PR TITLE
refactor: wallet management

### PR DESCRIPTION
**Pull request by klkvr**
## Motivation

This PR fixes two issues:
1. Users are forced to provide `--sender` argument when the sender can be determined from keystore or hw wallet. (#6034)
2. Users are forced to enter private keys and keystore password several times when running multi-chain scripts.

The idea was to rewrite `MultiWallet` into something that can be passed through script execution

## Solution

This PR consists of several parts:
1. Refactor of foundry-wallets. Initially the plan was to refactor it into 2 distinct parts: logic related to user input (clap, validation, etc) and logic related to actual signing and wallets, but it seems that besides the former logic foundry-wallets is basically 100 lines of thin wrappers around ethers-signers. Not sure what's the best way to deal with those, so right now I just extracted all logic for reading wallets and creating signers into reusable free functions in `wallets/utils`, so that it's easier to navigate and we no more duplicating stuff.
2. Add `MultiWallet` and `PendingWallet`. `MultiWallet` contains set of `WalletSigner` which is our enum wrapping ethers-signers objects and also a set of `PendingWallet`, those are wallets which require user action to be unlocked. Currently it's only keystores and interactively requested private keys. The idea here is that `MultiWallet` is an entity created once and used for the whole duration of the wallet interaction and on the first time signers are requested, all pending wallets are getting unlocked. 
3. To make signers independent from chain_id I changed our logic to completely ignore `chain_id` on signers and set it to 1 by default. This can be cleaned up once we migrate to alloy signers which have `chain_id` field optional
4. Add `MultiWallet` to cheatcodes inspector behing `Arc<Mutex>`. Use it in cheatcodes to determine sender for "anonymous" broadcasts and add remembered private keys directly to `MultiWallet` storage. Mutex here shouldn't really cause any threads to wait for locks as `Cheatcodes` are only getting cloned and used in several threads when running tests, for which `script_wallets` is being set to None as we are not collecting anything from it for tests.

Not sure if there was a demand for it, but this brings wallet-management closer to the cheatcodes, and would let us add support for `vm.sign` through local wallets (for example, signing stuff in script via ledger or keystore).

### CODE REVIEW

File: crates/cheatcodes/src/script.rs [hunk 1]
```diff
@@ -86,8 +96,25 @@ fn broadcast<DB: DatabaseExt>(
 
     correct_sender_nonce(ccx)?;
 
+    let mut new_origin = new_origin.cloned();
+
+    if new_origin.is_none() {
+        if let Some(script_wallets) = &ccx.state.script_wallets {
+            let mut script_wallets = script_wallets.lock().unwrap();
```
Comment (mattsse) [inline crates/cheatcodes/src/script.rs:103]: the Mutes should be `parking_lot::Mutex`

File: crates/cheatcodes/src/script.rs [hunk 2]
```diff
@@ -72,6 +73,15 @@ pub struct Broadcast {
...
 }
 
+/// Contains context for wallet management.
+#[derive(Debug)]
+pub struct ScriptWalletsData {
+    /// All signers in scope of the script.
+    pub multi_wallet: MultiWallet,
+    /// Optional signer provided as `--sender` flag.
+    pub provided_sender: Option<Address>,
+}
```
Comment (mattsse) [inline crates/cheatcodes/src/script.rs:83]: I see we're using `Arc<Mutex<ScriptWalletsData>` a few times,

can we have a wrapper type for this

```rust
struct ScriptWallets {
  // maybe Option?
  inner: Arc<Mutex<ScriptWallets>>
}
```

**Review by mattsse — CHANGES_REQUESTED** [review]
> but this brings wallet-management closer to the cheatcodes,

I think we want this, yes

style nits, 
ptal @DaniPopes

File: crates/cheatcodes/src/script.rs [hunk 3]
```diff
@@ -106,13 +159,16 @@ fn broadcast_key<DB: DatabaseExt>(
...
+    let new_origin = LocalWallet::from(key.clone()).address();
+
+    let result = broadcast(ccx, Some(&new_origin), single_call);
 
-    let result = broadcast(ccx, Some(new_origin), single_call);
     if result.is_ok() {
-        ccx.state.script_wallets.push(wallet);
+        let signer = WalletSigner::from_private_key(key.to_bytes())?;
+        if let Some(script_wallets) = &ccx.state.script_wallets {
+            script_wallets.inner.lock().multi_wallet.add_signer(signer);
```
Comment (mattsse) [inline crates/cheatcodes/src/script.rs:170]: this should be a function of the the scriptwallets type

File: crates/wallets/src/utils.rs [hunk 4]
```diff
@@ -0,0 +1,145 @@
+use ethers_signers::{HDPath as LedgerHDPath, LocalWallet, TrezorHDPath, WalletError};
+use foundry_config::Config;
+use std::{
+    fs,
+    path::{Path, PathBuf},
+    str::FromStr,
+};
+
+use eyre::{Context, Result};
+
```
Comment (mattsse) [inline crates/wallets/src/utils.rs:10]:
```suggestion
```

File: crates/wallets/src/utils.rs [hunk 5]
```diff
@@ -0,0 +1,145 @@
+use ethers_signers::{HDPath as LedgerHDPath, LocalWallet, TrezorHDPath, WalletError};
+use foundry_config::Config;
+use std::{
+    fs,
+    path::{Path, PathBuf},
+    str::FromStr,
+};
+
```
Comment (mattsse) [inline crates/wallets/src/utils.rs:8]:
```suggestion
```

File: crates/wallets/src/utils.rs [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+use eyre::{Context, Result};
+
+use crate::{error::PrivateKeyError, PendingSigner, WalletSigner};
+
+pub fn create_private_key_signer(private_key: &str) -> Result<WalletSigner> {
```
Comment (mattsse) [inline crates/wallets/src/utils.rs:13]: can this be a function on the WalletSigner type instead?

File: crates/wallets/src/wallet_signer.rs [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+
+pub type Result<T> = std::result::Result<T, WalletSignerError>;
+
+#[derive(Debug)]
+pub enum WalletSigner {
+    Local(LocalWallet),
+    Ledger(Ledger),
+    Trezor(Trezor),
+    Aws(AwsSigner),
+}
```
Comment (mattsse) [inline crates/wallets/src/wallet_signer.rs:28]: I'd appreciate some minimalistic docs here

File: crates/wallets/src/wallet_signer.rs [hunk 8]
```diff
@@ -0,0 +1,173 @@
+use std::path::PathBuf;
+
```
Comment (mattsse) [inline crates/wallets/src/wallet_signer.rs:2]:
```suggestion
```

File: crates/wallets/src/wallet_signer.rs [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+use ethers_signers::{
+    coins_bip39::English, AwsSigner, HDPath as LedgerHDPath, Ledger, LocalWallet, MnemonicBuilder,
+    Signer, Trezor, TrezorHDPath,
+};
+use rusoto_core::{
+    credential::ChainProvider as AwsChainProvider, region::Region as AwsRegion,
+    request::HttpClient as AwsHttpClient, Client as AwsClient,
+};
+use rusoto_kms::KmsClient;
+
```
Comment (mattsse) [inline crates/wallets/src/wallet_signer.rs:17]:
```suggestion
```

**Review by mattsse — CHANGES_REQUESTED** [review]
more nits, otherwise lgtm

pending @DaniPopes

File: crates/wallets/src/utils.rs (hunk 6 above)
Comment (klkvr) [reply crates/wallets/src/utils.rs:13]: yep, but it feels like this functions's logic doesn't really belong there tbh as it's more about user input validation then about WalletSigner construction

wdyt?

Comment (mattsse) [reply crates/wallets/src/utils.rs:11]: I see, actually keeping it in utils is fine then
a few docs would be great

File: crates/wallets/src/raw_wallet.rs [hunk 10]
```diff
@@ -37,3 +39,23 @@ pub struct RawWallet {
     #[clap(long, conflicts_with = "hd_path", default_value_t = 0, value_name = "INDEX")]
     pub mnemonic_index: u32,
 }
+
+impl RawWallet {
+    pub fn signer(&self) -> Result<Option<WalletSigner>> {
```
Comment (mattsse) [inline crates/wallets/src/raw_wallet.rs:44]: one line doc

**Review by mattsse — APPROVED** [review]
a few additional docs would be great, otherwise lgtm

pending @DaniPopes @Evalir

File: crates/evm/evm/src/executors/mod.rs [hunk 11]
```diff
@@ -12,7 +12,7 @@ use crate::inspectors::{
 use alloy_dyn_abi::{DynSolValue, FunctionExt, JsonAbiExt};
 use alloy_json_abi::Function;
 use alloy_primitives::{Address, Bytes, Log, U256};
-use alloy_signer::LocalWallet;
+
```
Comment (DaniPopes) [inline crates/evm/evm/src/executors/mod.rs:15]:
```suggestion
```

File: crates/forge/bin/cmd/script/cmd.rs [hunk 12]
```diff
@@ -198,8 +230,8 @@ impl ScriptArgs {
         &mut self,
         script_config: ScriptConfig,
         linker: Linker,
-        result: ScriptResult,
         mut verify: VerifyBundle,
+        signers: HashMap<Address, WalletSigner>,
```
Comment (DaniPopes) [inline crates/forge/bin/cmd/script/cmd.rs:234]:
```suggestion
        signers: &HashMap<Address, WalletSigner>,
```

File: crates/forge/bin/cmd/script/cmd.rs [hunk 13]
```diff
@@ -163,8 +195,8 @@ impl ScriptArgs {
         script_config: ScriptConfig,
         linker: Linker,
         libraries: Libraries,
-        result: ScriptResult,
         verify: VerifyBundle,
+        signers: HashMap<Address, WalletSigner>,
```
Comment (DaniPopes) [inline crates/forge/bin/cmd/script/cmd.rs:199]:
```suggestion
        signers: &HashMap<Address, WalletSigner>,
```

File: crates/forge/bin/cmd/script/multi.rs [hunk 14]
```diff
@@ -179,8 +181,8 @@ impl ScriptArgs {
         mut deployments: MultiChainSequence,
         libraries: Libraries,
         config: &Config,
-        script_wallets: Vec<LocalWallet>,
         verify: VerifyBundle,
+        signers: HashMap<Address, WalletSigner>,
```
Comment (DaniPopes) [inline crates/forge/bin/cmd/script/multi.rs:185]:
```suggestion
        signers: &HashMap<Address, WalletSigner>,
```

**Review by DaniPopes — APPROVED** [review]
lgtm, mostly just avoiding cloning

**Review by Evalir — APPROVED** [review]
lgtm, i really like this.

**Comment by rmcmk** [comment]
Hey, this is excellent. Nice.

**Comment by mattsse** [comment]
send it @klkvr ?

**Pull request merged**